[PR196 3/3] Adding support for partial workspace cleaning

jbohren opened the pull request:
This PR continues the refactoring of contributions from #196. It adds support for partial workspace cleaning, which takes advantage of the linked devel space pattern.

It varies slightly from the implementation in #196, and reorganizes some *under-the-hood* elements of `catkin_tools`:
- devel and install manifests are stored in `WS_ROOT/.catkin_tools/PROFILE/packages/PKG/`
- logs for each package are stored in `WS_ROOT/.catkin_tools/PROFILE/logs/PKG/`
- linked devel collisions file is stored in `WS_ROOT/.catkin_tools/PROFILE/`
- for linked devel, the *actual* isolated devel space is now referred to as the "private" devel space
- private devel spaces are stored in `DEVEL_SPACE/.private/PKG/`
- clean jobs are supplied by entry_points

Additionally, it includes the following changes:
- clean's `--dry-run` prints out each file / directory which is to be unlinked or removed
- adds `CATKIN_TOOLS_FORCE_COLOR` environment variable (this is to support #250)
- fixes alignment of catkin build `skipped` messages
- fixes counter-intuitive development bug where jobserver might not be initialized properly
- fixes bug with prebuild package scheduling

Tasks:
- [x] Reincorporate CLI options for controlling clean output
- [x] Add unit tests
- [x] Fix #288 
- [x] Fix #212
- [x] move profiles into `profiles` metadata subdir
- [x] update metadata subdir README
- [x] change `clean --deps` option to `--dependencies`
- [x] `catkin clean catkin` doesn't work -- build catkin into private devel space
- [x]  make `--deinit` respect `--dry-run` (and other actions)
- [x] make `--dry-run` clearer
- [x] remove superfluous `--all` option 
- [x] fix `--orphans` bug
- [x] move logs into first-class "log space"

- [x] reduce verbosity of warnings for already cleaned objects
- [x] make space cleaning and package cleaning mutually exclusive
- [x] always force cmake to run after cleaning
- [x] add interactive checks for space cleaning
- [x] make output from different cleaning stages more uniform
- [x] make bare `clean` also clean logs
- [x] make `--deinit` also clean spaces
- [x] fix bug that prevents changing the devel layout before the first build
- [x] fix catkin prebuild https://github.com/catkin/catkin_tools/pull/293#issuecomment-199963925

Full CLI interface:

```
usage: catkin clean [-h] [--workspace WORKSPACE] [--profile PROFILE]
                    [--dry-run] [--verbose] [--force] [-a] [-b] [-d] [-i]
                    [--deinit] [--deps] [--orphans] [-l] [--setup-files]
                    [PKGNAME [PKGNAME ...]]

Deletes various products of the build verb.

optional arguments:
  -h, --help            show this help message and exit
  --workspace WORKSPACE, -w WORKSPACE
                        The path to the catkin_tools workspace or a directory
                        contained within it (default: ".")
  --profile PROFILE     The name of a config profile to use (default: active
                        profile)
  --dry-run, -n         Show the effects of the clean action without modifying
                        the workspace.
  --verbose, -v         Verbose status output.
  --force, -f           Skip all interactive checks.

Basic:
  Clean workspace subdirectories.

  -a, --all             Remove all of the generated spaces associated with the
                        given or active profile. This will remove everything
                        but the source space and the hidden .catkin_tools
                        directory.
  -b, --build           Remove the entire buildspace.
  -d, --devel           Remove the entire develspace.
  -i, --install         Remove the entire installspace.
  --deinit              De-initialize the workspace, delete all build profiles
                        and configuration.

Packages:
  Clean products from specific packages in the workspace. These options will
  automatically enable the --force-cmake option for the next build
  invocation.

  PKGNAME               Explicilty specify a list of specific packages to
                        clean from the build, devel, and install space.
  --deps                Clean the packages which depend on other packages to
                        be cleaned.
  --orphans             Remove products from packages are no longer in the
                        source space. Note that this also removes packages
                        which are blacklisted or which contain `CATKIN_INGORE`
                        marker files.

Advanced:
  Clean only specific parts of the workspace for specified packages. These
  options will automatically enable the --force-cmake option for the next
  build invocation.

  -l, --logs            Only clear the catkin-generated logfiles.
  --setup-files         Clear the catkin-generated setup files from the devel
                        and install spaces.
```

wjwwood commented:
@jbohren I have some questions about the design goals stated in the OP, I'll try to address them inline:

> devel and install manifests are stored in `WS_ROOT/.catkin_tools/PROFILE/packages/PKG/`

These are files with a list of files installed to either the devel space or install space respectively or are they `package.xml` package manifests?

If the former, then why are they in a subfolder `packages/<package name>/` rather than something like `manifests/<package name>/` or `install_manifests/<package name>.txt` and `devel_manifests/<package name>.txt`?

> logs for each package are stored in WS_ROOT/.catkin_tools/PROFILE/logs/PKG/

I assume this means the build logs? Why move them from the build space? (I thought the build space was quite an appropriate place for the logs)

> linked devel collisions file is stored in `WS_ROOT/.catkin_tools/PROFILE/`

Is that in a single file or some other organization scheme?

wjwwood commented:
> for linked devel, the actual isolated devel space is now referred to as the "private" devel space
> private devel spaces are stored in `DEVEL_SPACE/.private/PKG/`

Why not fall back to the default `catkin` behavior an let the devel space for each package be created in the package's build space, i.e. `build/<package name>/devel`?

wjwwood commented:
> clean jobs are supplied by entry_points

Can you elaborate on this more (bullet list of steps for the process)? You mean Python `entry_points`?

wjwwood commented:
> adds `CATKIN_TOOLS_FORCE_COLOR` environment variable (this is to support #250)

Having not looked at #250 in detail, why do you need an environment variable to do this? Can you not just use `catkin --force-color ...`?

wjwwood commented:
That's all the questions I have so far. I'm going to start using it and testing it out locally.

Btw, this needs to be rebased against master to be mergeable.

jbohren commented:
> @jbohren I have some questions about the design goals stated in the OP, I'll try to address them inline:

>> devel and install manifests are stored in `WS_ROOT/.catkin_tools/PROFILE/packages/PKG/`

> These are files with a list of files installed to either the devel space or install space respectively or are they `package.xml` package manifests?

> If the former, then why are they in a subfolder `packages/<package name>/` rather than something like `manifests/<package name>/` or `install_manifests/<package name>.txt` and `devel_manifests/<package name>.txt`?

It is the former, these are like CMake `install_manifest.txt` files. I chose `packages` as the  subdir name, but `manifests` or `product_manifests` could also be appropriate, but it might get cluttered if we wanted to store other package-specific information in there. For example, in the future there might be package-specific config defaults.

---

>> logs for each package are stored in WS_ROOT/.catkin_tools/PROFILE/logs/PKG/

> I assume this means the build logs? Why move them from the build space? (I thought the build space was quite an appropriate place for the logs)

The build space was an appropriate place for the logs when the `build` verb was the only verb generating logs. With the more advanced `clean` verb, there are also `clean` logs. Additionally, a user might want logs to persist even after clearing the build space, so this puts them in a safer place.

---

>> linked devel collisions file is stored in `WS_ROOT/.catkin_tools/PROFILE/`

> Is that in a single file or some other organization scheme?

That's a single file called `devel_collisions.txt` which lists files and their collision counts. This is a locked (synchronized) resource.

---

>> for linked devel, the actual isolated devel space is now referred to as the "private" devel space
>> private devel spaces are stored in `DEVEL_SPACE/.private/PKG/`

> Why not fall back to the default `catkin` behavior an let the devel space for each package be created in the package's build space, i.e. `build/<package name>/devel`?

That could also be an appropriate place. It depends on whether you want the devel space to continue to work if the build space is removed.

---

> > adds `CATKIN_TOOLS_FORCE_COLOR` environment variable (this is to support #250)

> Having not looked at #250 in detail, why do you need an environment variable to do this? Can you not just use `catkin --force-color ...`?

This is used to generate the asciicasts for the new documetation with colored output without having to put `--force-color` in front of all the commands: https://github.com/catkin/catkin_tools/blob/pre-0.4.0-docs/docs/examples/slowrun#L55

e.g.:
[![asciicast](https://asciinema.org/a/7tjl75sm7lbp7jlgc5tlrsps6.png)](https://asciinema.org/a/7tjl75sm7lbp7jlgc5tlrsps6)

wjwwood commented:
> It is the former, these are like CMake install_manifest.txt files. I chose packages as the subdir name, but manifests or product_manifests could also be appropriate, but it might get cluttered if we wanted to store other package-specific information in there. For example, in the future there might be package-specific config defaults.

If the plan is to potentially put other package specific stuff there in the future, then that's probably ok.

> The build space was an appropriate place for the logs when the build verb was the only verb generating logs. With the more advanced clean verb, there are also clean logs. Additionally, a user might want logs to persist even after clearing the build space, so this puts them in a safer place.

The clean logs use case is a compelling reason to not put them in the build space, but I worry that putting  them in the `.catkin_tools` directory will mean they never get cleaned and just build up over time. Does the `-l` option to `catkin clean` clean the clean logs too (:dizzy_face:)?

What if we just put logs in a top level logs folder like we have for build, devel and install? It could still contain a folder hierarchy with profile names in the mix. For example:

```
% tree -L 2
.
├── build
│   ├── actionlib
│   ├── ...
│   └── xmlrpcpp
├── desktop-indigo.rosinstall
├── devel
│   ├── bin
│   ├── ...
│   └── share
├── install
│   ├── _setup_util.py
│   ├── ...
│   └── share
├── logs
│   ├── actionlib
│   ├── ...
│   └── xmlrpcpp
└── src
    ├── actionlib
    ├── ...
    └── xacro
```

And then either logs can get a suffix for each profile (like the build, devel, etc.) or it can be subdivided internally by profile and then package?

> That could also be an appropriate place. It depends on whether you want the devel space to continue to work if the build space is removed.

Is that common? Maybe, but they could always specify the devel space root if they wanted it out of build. I'm not sure what the preferable default case would be, but I'm leaning towards the build space since it doesn't introduce a new folder and concept into the mix.

> The job factories are registered along with build job factories like so:
https://github.com/catkin/catkin_tools/blob/pre-0.4.0-clean-pkgs/catkin_tools/jobs/catkin.py#L528-L533

> These are gathered in the clean verb here:
https://github.com/catkin/catkin_tools/blob/pre-0.4.0-clean-pkgs/catkin_tools/verbs/catkin_clean/clean.py#L124-L127

I see, I thought you meant the actual instances of jobs, you just mean the specialization of clean for a particular build type.

> This is used to generate the asciicasts for the new documetation with colored output without having to put --force-color in front of all the commands: https://github.com/catkin/catkin_tools/blob/pre-0.4.0-docs/docs/examples/slowrun#L55

Ok, so long as there was a reason. Sounds good to me.

Thanks for answering all my questions.

wjwwood commented:
@jbohren I rebased against master and force pushed.

wjwwood commented:
Some usability notes:

- It's not clear from the documentation what the default behavior of `catkin clean` is with no extra arguments.
- It's not clear from the console output that a dry run is happening. Perhaps there should be a message when `-n` is used.
- When cleaning a space which doesn't exist, there is no output.
  - I'm not sure if this is the right thing to do or not, but my initial reaction was that it should say it would have removed the install space (for example) but it does not exist.
- `--deinit` doesn't seem to respect `--dry-run`.
- `catkin clean -n catkin` doesn't seem to work, or rather `catkin` doesn't seem to have a private build space.
- I'd recommend change the `--deps` option to be `--dependents` and `-d` for short, as `--deps` is commonly referred to as `--dependencies`.
- Should there also be a `--dependencies` or `--unique-dependencies`? Like if I wanted to remove a package and all of its dependencies so long as another package does not use those dependencies? It's sort of like `apt-get`'s `autoremove` but without the notion of selected versus implicit.

wjwwood commented:
I just ran into this problem:

```
% catkin clean -n rviz --orphans
[clean] Determining orphaned packages...
[clean] Orphaned build products: .built_by
[clean] Note: Parts of the workspace have been cleaned which will necessitate re-configuring CMake on the next build.
Traceback (most recent call last):
  File "/usr/local/bin/catkin", line 9, in <module>
    load_entry_point('catkin-tools==0.3.1', 'console_scripts', 'catkin')()
  File "/Library/Python/2.7/site-packages/catkin_tools/commands/catkin.py", line 265, in main
    catkin_main(sysargs)
  File "/Library/Python/2.7/site-packages/catkin_tools/commands/catkin.py", line 260, in catkin_main
    sys.exit(args.main(args) or 0)
  File "/Library/Python/2.7/site-packages/catkin_tools/verbs/catkin_clean/cli.py", line 245, in main
    for pkg_devel_name in ctx.private_devel_path():
TypeError: 'str' object is not callable
```

The scenario was that I had an indigo desktop full workspace with everything built and configured to do the linked devel space.

jbohren commented:
>- It's not clear from the documentation what the default behavior of `catkin clean` is with no extra arguments.

Yeah, we should get rid of the `--all` option, since that's the default behavior.

>- It's not clear from the console output that a dry run is happening. Perhaps there should be a message when `-n` is used.

+1

>- When cleaning a space which doesn't exist, there is no output.
>  - I'm not sure if this is the right thing to do or not, but my initial reaction was that it should say it would have removed the install space (for example) but it does not exist.

+1

>- `--deinit` doesn't seem to respect `--dry-run`.

And it should.

>- `catkin clean -n catkin` doesn't seem to work, or rather `catkin` doesn't seem to have a private build space.

Yeah, currently prebuild packages are built directly into the devel space, but really they should be linked just like the others.

>- I'd recommend change the `--deps` option to be `--dependents` and `-d` for short, as `--deps` is commonly referred to as `--dependencies`.
catkin is 
+1

>- Should there also be a `--dependencies` or `--unique-dependencies`? Like if I wanted to remove a package and all of its dependencies so long as another package does not use those dependencies? It's sort of like `apt-get`'s `autoremove` but without the notion of selected versus implicit.

Maybe in the future.

NikolausDemmel commented:
I have been using this for a few weeks and have a few comments. Otherwise as always: Nice stuff! Thanks!

> For example, in the future there might be package-specific config defaults.

+1 for that feature (edit: this enhancement is described in #205)

>>  for linked devel, the actual isolated devel space is now referred to as the "private" devel space
>>  private devel spaces are stored in DEVEL_SPACE/.private/PKG/

> Why not fall back to the default catkin behavior an let the devel space for each package be created in the package's build space, i.e. build/<package name>/devel?

Not sure if there is a use case of wiping build yet keeping devel, but I would find it more logical if the devel space would be self-contained (together with the source space) and not depend on files that are actually in the build space, in particular if that is the case now (not sure if that is currently the case or not). Changing the devel products to be in the build space would also make the result of the two commands `catkin clean --devel` and `catkin clean --build` less intuitive, since `--build` would essentially inply `--devel`, at least with the current implementation.

> Should there also be a --dependencies or --unique-dependencies? Like if I wanted to remove a package and all of its dependencies so long as another package does not use those dependencies? It's sort of like apt-get's autoremove but without the notion of selected versus implicit.

I can see that use case, but there is also a use case for cleaning a package and all its dependencies (even non-unique ones), so one should be careful with the naming. `--unique-dependencies` for your use case and `--dependencies` for the one I am describing could make sense. Out of scope for this PR, so I opened a new enhancement issue 

> When cleaning a space which doesn't exist, there is no output.

>    I'm not sure if this is the right thing to do or not, but my initial reaction was that it should say it would have removed the install space (for example) but it does not exist.

Maybe there should be an output, but no warning.

Moreover, if you call `catkin clean foo` on a package that is already cleaned (or not yet built i guess), you get 3 warnings like

```
$ catkin clean foo
Starting  >>> foo   
_________________________________________________________________________________________________________________________________________________________________________________________________________________________________
Warnings   << foo:unlink /home/den2pal/work/ros_ws/.catkin_tools/default/logs/foo/clean.unlink.009.log
Warning: No private devel path found at `/home/den2pal/work/ros_ws/devel/.private/foo`
.................................................................................................................................................................................................................................
_________________________________________________________________________________________________________________________________________________________________________________________________________________________________
Warnings   << foo:rmdevel /home/den2pal/work/ros_ws/.catkin_tools/default/logs/foo/clean.rmdevel.009.log
Warning: File /home/den2pal/work/ros_ws/devel/.private/foo could not be deleted because it does not exist.
.................................................................................................................................................................................................................................
_________________________________________________________________________________________________________________________________________________________________________________________________________________________________
Warnings   << foo:rmbuild /home/den2pal/work/ros_ws/.catkin_tools/default/logs/foo/clean.rmbuild.009.log
Warning: File /home/den2pal/work/ros_ws/build/foo could not be deleted because it does not exist.
.................................................................................................................................................................................................................................
Finished  <<< foo              [ 0.0 seconds ]

```

Moreover, I noticed the following issues single package base cleaning:

1. The following is surprising in two ways: Firstly, it actually removes the whole devel space (where I would have expected it to only remove the devel space of `foo`). Secondly it gives a strange error message, since my devel space configuration is set to `linked`. It is the same with `--build`, or `--all`.

```
$ catkin clean foo -n --devel
[clean] Removing develspace: /home/den2pal/work/ros_ws/devel
[clean] Error: Individual packages can only be cleaned from workspaces with symbolically-linked develspaces (`catkin config --link-devel`).
``

2. The logic behind `Note: Workspace packages have changed, please re-source setup files to use them.` does only seem to work if you build the entire workspace after cleaning a individual packages. We would probably need to store a "force cmake on next run" for each package individually. For example consider `a_pkg` and `b_pkg` that depends on `a_pkg` in a fully built workspace. If I now `catkin clean a_pkg` and then `catkin build a_pkg`, a subsequent `catkin build b_pkg` does not currently invoke cmake for `b_pkg`, but it probably should, no?

jbohren commented:
@wjwwood This is good to review, all outstanding issues are resolved as of #09e2b12. I've also made the context config a little more consistent / succinct. I'll be making another bunch of commits to the doc PR to capture the changes.

wjwwood commented:
Looks like Python3 is failing on all platforms.

jbohren commented:
@wjwwood It looks like in the last month something has changed with the Trollius / Python3 interface so the executor no longer works. Is this related to the changes in osrf_pycommon?

NikolausDemmel commented:
I just noticed that when upgrading to the latest version on this branch, not all parts of my config got migrated (or possibly nothing)? In any case before this gets released we might need to think about a configuration migration scheme, or at least inform the user that they need to reconfigure their catkin tools workspaces.

wjwwood commented:
I suppose it could be, but just looking at the code I think what you're doing will not work in asyncio. Perhaps you were using trollius even in Python3 up until now?

```
  File "/home/travis/build/catkin/catkin_tools/catkin_tools/execution/executor.py", line 281, in execute_jobs
    return_when=FIRST_COMPLETED))
RuntimeError: yield was used instead of yield from for generator in task <Task pending coro=<execute_jobs() running at /home/travis/build/catkin/catkin_tools/catkin_tools/execution/executor.py:281> cb=[_run_until_complete_cb() at /opt/python/3.4.3/lib/python3.4/asyncio/base_events.py:123]> with <generator object wait at 0x7fce5a10efc0>
```

From: https://github.com/catkin/catkin_tools/blob/2689f63d4c7495bb390d2c909c1871e2c1b21bb1/catkin_tools/execution/executor.py#L24
```
import trollius as asyncio
```

From: https://github.com/catkin/catkin_tools/blob/2689f63d4c7495bb390d2c909c1871e2c1b21bb1/catkin_tools/execution/executor.py#L278-L281
```
        done_job_fs, active_job_fs = yield asyncio.From(asyncio.wait(
            active_job_fs,
            timeout=0.10,
            return_when=FIRST_COMPLETED))
```

The above doesn't appear to use `osrf_pycommon` directly, though I suppose it could be that `osrf_pycommon` could be contributing to the contents of `active_job_fs` indirectly?

wjwwood commented:
It might also be an interaction between asyncio and trollius being used simultaneously, which is at least possible. Trollius 2.1 was released in february (https://pypi.python.org/pypi/trollius), so maybe you were only using 2.0 at the time? Not sure what the issue could be.

wjwwood commented:
Double check which system the loop is from, it could be that `osrf_pycommon` is giving you an asyncio loop but you're trying to use it with trollius. We may need a way to tell `osrf_pycommon` which system to use.

NikolausDemmel commented:
With 'linked' develspace and a fresh workspace / config I get:

```
$ catkin build
Error: The current devel space layout, `merged`,is incompatible with the configured layout, `linked`.
```


Other things I noticed:

## with 'merged' develspace (not sure if it makes a difference for these issues):
* `catkin clean --build` does not ask interactively any more; same with --devel, --log, ... (only --deinit seems to have an interactive check)
* output is a bit inconsistent (in the first example it does not mention build not existing, in the second, it mentions log does not exist):
```
$ catkin build
$ rm -r build/
$ catkin clean -dlb
[clean] Removing develspace: /tmp/foo/devel
[clean] Removing log files from: /tmp/foo/logs
```

```
$ catkin build
$ rm -r logs/
$ catkin clean -dlb
[clean] Removing develspace: /tmp/foo/devel
[clean] Removing buildspace: /tmp/foo/build
[clean] Log file directory does not exist: /tmp/foo/logs
```

NikolausDemmel commented:
Moreover, to me it seems problematic that files like the `build.yaml` is not in the build space and files like the `devel_collisions` and `devel_manifest` are not in the devel space.

If I manually remove `build` or `devel`, I will not get the "fresh start" like I used to before. Does `catkin clean` actually take care to remove these files?

To give a very concrete example with the `build.yaml`:

```
catkin init
catkin build
catkin clean -dbl
catkin config --link-devel
catkin build
# Here catkin complains that the last devel layout was 'merged' and thus refuses to build
# You can currently fix it only by manually removing or editing '.catkin_tools/profiles/default/build.yaml'
```

NikolausDemmel commented:
One more thing actually (this is already the case in the version of this branch before the updates today):

If you build a workspace without `catkin` in in the source space, it builds `catkin_tools_prebuild` exactly once, and never rebuilds it. That seems fine. If you have `catkin` checked out in the source space, it builds it every time, even if you call `catkin build foo-pkg --no-deps`. Is that needed?

jbohren commented:
@wjwwood 
> Double check which system the loop is from, it could be that osrf_pycommon is giving you an asyncio loop but you're trying to use it with trollius. We may need a way to tell osrf_pycommon which system to use.

Yeah, that's probably it. I'll look at that later tonight. If you re-run any of the old Travis builds, you'll see that they fail now on Python3, even if they succeeded before.

@NikolausDemmel You've been giving a whole bunch of good feedback! I'll enumerate / respond to it a little later, as well.

jbohren commented:
@wjwwood Since https://github.com/osrf/osrf_pycommon/commit/8997085d8abbfb56369d56112fa45fe201a9c303, `osrf_pycommon` no longer imports trollius on Python 3.4 even if you import it explicitly. I thought we were sticking with trollius on Python3 for the time being, did something change or is this in error?

wjwwood commented:
@jbohren it's fine for `catkin_tools` to depend on Trollius at all versions of Python, but for osrf_pycommon, it would be best if it could avoid Trollius if/when asyncio is available. I guess we will need to add an option to `osrf_pycommon` so that you can force it to use a trollius loop. This is the key sentence from the Trollius documentation:

From: http://trollius.readthedocs.org/asyncio.html#write-code-working-on-trollius-and-asyncio
> Since Trollius 0.4, it’s possible to use asyncio and Trollius coroutines in the same process. The only limit is that the event loop must be a Trollius event loop.

wjwwood commented:
I'll look into that tonight.

jbohren commented:
I modified the travis script to use `osrf_pycommon` 0.1.0, and explicitly import the trollius functions, and it's passing on Python 3.4 (OS X still pending): https://travis-ci.org/catkin/catkin_tools/builds/117822048

jbohren commented:
@NikolausDemmel Let's see if I can cover everything...

---

>> For example, in the future there might be package-specific config defaults.
>+1 for that feature (edit: this enhancement is described in #205)

Yeah I think this is definitely something that we can look at with 0.5.0.

---

> Not sure if there is a use case of wiping build yet keeping devel, but I would find it more logical if the devel space would be self-contained (together with the source space) and not depend on files that are actually in the build space

Yeah, I think we're in agreement that "devel stuff" should be in the devel space, even if it's in the `.private` directory.

---

> I can see that use case, but there is also a use case for cleaning a package and all its dependencies (even non-unique ones), so one should be careful with the naming. --unique-dependencies for your use case and --dependencies for the one I am describing could make sense. Out of scope for this PR, so I opened a new enhancement issue #302

Yeah, right now we just have `--dependants`, which means cleaning all the packages which *depend on* a given package. We can discuss other functionality in #302.

---

>>When cleaning a space which doesn't exist, there is no output.

> Moreover, if you call catkin clean foo on a package that is already cleaned (or not yet built i guess), you get 3 warnings like

Yeah, I think these warnings should instead just be normal output, and should say something like `File ... has already been deleted.`

---

> The following is surprising in two ways: Firstly, it actually removes the whole devel space (where I would have expected it to only remove the devel space of foo). Secondly it gives a strange error message, since my devel space configuration is set to linked. It is the same with --build, or --all.

> ```
$ catkin clean foo -n --devel
[clean] Removing develspace: /home/den2pal/work/ros_ws/devel
[clean] Error: Individual packages can only be cleaned from workspaces with symbolically-linked develspaces (`catkin config --link-devel`).
```

Yeah, I think the interface should be simplified here. I think you should either be able to clean some set of spaces OR some set of packages, but I don't see enough utility in cleaning a package from only one of the spaces.

---

> The logic behind `Note: Workspace packages have changed, please re-source setup files to use them.` does only seem to work if you build the entire workspace after cleaning a individual packages. We would probably need to store a "force cmake on next run" for each package individually. For example consider a_pkg and b_pkg that depends on a_pkg in a fully built workspace. If I now catkin clean a_pkg and then catkin build a_pkg, a subsequent catkin build b_pkg does not currently invoke cmake for b_pkg, but it probably should, no?

It would definitely be better to only force a subset of packages to reconfigure, but for now a global force-cmake should work.

---

I just noticed that when upgrading to the latest version on this branch, not all parts of my config got migrated (or possibly nothing)? In any case before this gets released we might need to think about a configuration migration scheme, or at least inform the user that they need to reconfigure their catkin tools workspaces.

Yeah, this should have automatic migration before it gets merged in.

---

> With 'linked' develspace and a fresh workspace / config I get:

>```
$ catkin build
Error: The current devel space layout, `merged`,is incompatible with the configured layout, `linked`.
```

Yeah this is a bug.

---

>catkin clean --build does not ask interactively any more; same with --devel, --log, ... (only --deinit seems to have an interactive check)

Yeah I think this is an oversight.

---

>output is a bit inconsistent (in the first example it does not mention build not existing, in the second, it mentions log does not exist):

Yeah this should be cleaned up.

---

> Moreover, to me it seems problematic that files like the `build.yaml` is not in the build space and files like the `devel_collisions` and `devel_manifest` are not in the devel space.

There's a bug here, but I don't think this is inherently problematic. What it should do is detect when these directories have been removed, and ignore the metadata state when that happens. We're already testing for existence of the spaces, so this isn't hard to add.

Since cleaning adds a whole new dimension to maintaining workspace state, I think it's worth adding a "workspace trouble finder" verb which does a more in-depth consistency check in case someone has been mucking around in the build/devel spaces manually.

NikolausDemmel commented:
@jbohren, thanks for taking the time to respond. I'll reply to some items where we don't yet have consensus.

> @NikolausDemmel Let's see if I can cover everything...

I think you replied to all my input, except for [the very last comment](https://github.com/catkin/catkin_tools/pull/293#issuecomment-199963925).

---

> Yeah, I think the interface should be simplified here. I think you should either be able to clean some set of spaces OR some set of packages, but I don't see enough utility in cleaning a package from only one of the spaces.

That is fine by me.

---

>> The logic behind Note: Workspace packages have changed, please re-source setup files to use them. does only seem to work if you build the entire workspace after cleaning a individual packages. We would probably need to store a "force cmake on next run" for each package individually. For example consider a_pkg and b_pkg that depends on a_pkg in a fully built workspace. If I now catkin clean a_pkg and then catkin build a_pkg, a subsequent catkin build b_pkg does not currently invoke cmake for b_pkg, but it probably should, no?

> It would definitely be better to only force a subset of packages to reconfigure, but for now a global force-cmake should work.

Forcing only a subset of packages to reconfigure might be a nice enhancement (albeit not really very important IMHO), but that is not actually what I was talking about. My concern is with when you mark all packages to reconfigure, but then only build a subset of packages in the next `catkin build` invocation. With the current global flag, all packages that were not built in that invocation, will not know that they should also reconfigure the next time they are built. I think the only solution is a per-package flag.

---

>> Moreover, to me it seems problematic that files like the build.yaml is not in the build space and files like the devel_collisions and devel_manifest are not in the devel space.

> There's a bug here, but I don't think this is inherently problematic. What it should do is detect when these directories have been removed, and ignore the metadata state when that happens. We're already testing for existence of the spaces, so this isn't hard to add.

Sounds reasonable.

> Since cleaning adds a whole new dimension to maintaining workspace state, I think it's worth adding a "workspace trouble finder" verb which does a more in-depth consistency check in case someone has been mucking around in the build/devel spaces manually.

How about `catkin doctor`, akin to `brew doctor`?

wjwwood commented:
CI is passing again with https://github.com/osrf/osrf_pycommon/pull/22. I'll merge that and release `osrf_pycommon` again.

jbohren commented:
@NikolausDemmel I enumerated all of the action items from your feedback above, and yeah, the catkin prebuild thing needs to be fixed. 

@wjwwood I'm going to implement these fixes now, and then it should be good to merged.

---

>> Since cleaning adds a whole new dimension to maintaining workspace state, I think it's worth adding a "workspace trouble finder" verb which does a more in-depth consistency check in case someone has been mucking around in the build/devel spaces manually.
> How about catkin doctor, akin to brew doctor?

I don't know, I really think "workspace trouble finder" has a nice ring to it...

NikolausDemmel commented:
@jbohren thanks!

>>> Since cleaning adds a whole new dimension to maintaining workspace state, I think it's worth adding a "workspace trouble finder" verb which does a more in-depth consistency check in case someone has been mucking around in the build/devel spaces manually.
>> How about catkin doctor, akin to brew doctor?
> I don't know, I really think "workspace trouble finder" has a nice ring to it...

True. Another idea in good ROS tradition is `catkin wtf`.

jbohren commented:
>>>> Since cleaning adds a whole new dimension to maintaining workspace state, I think it's worth adding a "workspace trouble finder" verb which does a more in-depth consistency check in case someone has been mucking around in the build/devel spaces manually.

>>> How about catkin doctor, akin to brew doctor?

>> I don't know, I really think "workspace trouble finder" has a nice ring to it...

> True. Another idea in good ROS tradition is `catkin wtf`.

:sunglasses:

NikolausDemmel commented:
It just struck me, yes...

davetcoleman commented:
:+1:

wjwwood commented:
Clearly it is `caktin where's the fire?`. :smile:

jbohren commented:
@NikolausDemmel 7bb6553 should fix all the issues you found. Thanks for the testing! The biggest issue was the interplay between the `catkin` and `catkin_tools_prebuild` package, and when someone adds or removes the `catkin` package from their workspace.

One thing I didn't find was any configuration migration issues. Which versions of `catkin_tools` were you switching between?

jbohren commented:
@wjwwood When does it make sense to make `linked` the default devel layout?

NikolausDemmel commented:
Cool! I'll give it a try next week.

The migration issue was updating from an earlier version of this branch. I think the main issue might have been that the profiles were moved into a subdirectory "profiles". I'll have to check next week.

+1 from me for making linked devel the default. I already use Individual package cleaning all the time with the earlier Version of this branch.

NikolausDemmel commented:
PS: merged or isolated should be used when using this on an existing devel space.

jbohren commented:
> The migration issue was updating from an earlier version of this branch. I think the main issue might have been that the profiles were moved into a subdirectory "profiles". I'll have to check next week.

Right yeah, that will do it.

jbohren commented:
> The migration issue was updating from an earlier version of this branch. I think the main issue might have been that the profiles were moved into a subdirectory "profiles". I'll have to check next week.

@NikolausDemmel Migration from `0.3.1` to `pre-0.4.0` works now with cf5fcda, but I haven't tested it with other `pre-0.4.0` versions

wjwwood commented:
I suppose it would make sense to make it the default asap so people can take advantage of the package cleaning.

In my opinion, it would also be ideal to start moving away from the devel space (defaulting to install) because there are so many issues with it, but that's a more drastic change.

jbohren commented:
> I suppose it would make sense to make [linked devel] the default asap so people can take advantage of the package cleaning.

Sounds good to me. :thumbsup:

> In my opinion, it would also be ideal to start moving away from the devel space (defaulting to install) because there are so many issues with it, but that's a more drastic change.

Yeah, the biggest hurdle there is migrating numerous packages out in the community without install targets. Fortunately, we can use the linked-devel architecture to help people make sure they're installing everything they need.

NikolausDemmel commented:
@jbohren, thanks for the update! It is looking really good now. I will do some more testing over the next few days, but here is some initial feedback (only minor things so far):

--- 

Workspace migration did something for me, but not quite what was needed. This was from a previous version of this branch as I said. I didn't investigate more, but I think the following happened (I only have the default profile in this workspace):

1. The default profile folder got moved into the subdirectory.
2. The `build.yaml` seemed to be correct
3. The `config.yaml` seemed to contain default information only, i.e. my customizations were lost. I was able to manually recover them from the `build.yaml`.
4. The logs still in the profile folder. They did not get cleaned from there with `catkin clean -l`.

I now manually removed the logs, reconfigured the default profile to my needs, cleaned the whole workspace, and I am now building anew.

I have no clue how much of the issues are due to me not being on 0.3.1, but some intermediate version. To me it is reasonable if migration is only supported from the last release, so maybe no action needed.

---

Just a very minor output inconsistency with `catkin clean`:

If none of the build-, devel-, log- dirs exists (probably install as well, but I haven't tried), I get a message:
```
$ catkin clean -b
[clean] No subdirectoris for the `default` profile exist. Nothing to be done.
```

If at least one exists (no matter which), I get no message:
```
$ catkin clean -b
```

I would remove the message in the former case.

wjwwood commented on catkin_tools/verbs/catkin_clean/cli.py:
```diff
@@ -14,178 +14,367 @@
+                        opts.verbose,
+                        opts.dry_run)
+                except KeyboardInterrupt:
+                    wide_log("[build] User interrupted!")
```
This hasn't been imported, I guess you just need to do `from catkin_tools.common import wide_log`.

wjwwood commented on catkin_tools/verbs/catkin_clean/cli.py:
```diff
@@ -14,178 +14,367 @@
+                        opts.dry_run)
+                except KeyboardInterrupt:
+                    wide_log("[build] User interrupted!")
+                    event_queue.put(None)
```
`event_queue` is not defined here, perhaps you forgot to pass it when moving this code into a function?

wjwwood commented on catkin_tools/verbs/catkin_clean/cli.py:
```diff
@@ -14,178 +14,367 @@
+
+    # Check for all profiles option
+    if opts.all_profiles:
+        profiles = get_profile_names(workspace_path)
```
Neither `get_profile_names` nor `workspace_path` is defined here.

wjwwood commented on catkin_tools/jobs/catkin.py:
```diff
@@ -430,73 +485,99 @@ def create_catkin_build_job(context, package, package_path, dependencies, force_
+            # Remove build targets from devel space
+            stages.append(CommandStage(
+                'clean',
+                [MAKE_EXEC, 'clean'] + make_args,
```
`make_args` is not defined here, should it just be removed?

wjwwood commented:
I made a few clean up commits.

wjwwood commented on catkin_tools/verbs/catkin_clean/clean.py:
```diff
@@ -0,0 +1,201 @@
+        all_succeeded = run_until_complete(ej)
+    except Exception:
+        status_thread.keep_running = False
+        all_succeeded = False
```
`all_succeeded` doesn't appear to be used. Should it be? If not please remove it.

wjwwood commented on catkin_tools/verbs/catkin_clean/clean.py:
```diff
@@ -0,0 +1,201 @@
+    :returns: full list of package names to be cleaned
+    :rtype: list
+    """
+    start = time.time()
```
The `start` is collected, but never used. Should this be removed?

wjwwood commented on catkin_tools/verbs/catkin_build/build.py:
```diff
@@ -375,34 +372,79 @@ def build_isolated_workspace(
     prebuild_jobs = {}
-    setup_util_exists = os.path.exists(os.path.join(context.devel_space_abs, '_setup_util.py'))
-    if context.link_devel and (not setup_util_exists or (force_cmake and len(packages) == 0)):
+    no_packages = len(packages) == 0
```
This doesn't appear to be used, should it be? If not please remove it.

wjwwood commented:
@jbohren the tool seems to work as expected to me. I think it would be best if you could comment on the few places I mentioned something in line and then merge this. We can follow up with new problems and UX problems in new issues against the master branch as we prepare for the 0.4 release.

jbohren commented:
@wjwwood I resolved the unused / missing elements you mentioned, and fixed the inconsistency that @NikolausDemmel mentioned.

Something else I did was replace the `--force` command with `--yes` since it better matches the relationship to the interactive prompt. Additionally, I realized that some users might try to shoot themselves in the foot and set up the installspace / destdir to be their root prefix. In this case I added a check (which can be overridden by `--force`) which prompts the user if any of the spaces to be removed are outside of the workspace root.

This should be good to merge, I'll update the documentation, and then I'll draft a call for testing / RFC that we can send out to ros-users.

wjwwood commented:
@jbohren Thanks for the quick turn around. The changes look good to me, but the CI is failing. I'll test out the new CLI options in the meantime. If things look ok (and CI is passing) I'll merge.

jbohren commented:
@wjwwood Yeah, I moved the cli check to the wrong place. 13aa6e5 passes locally, and should pass CI now.

jbohren commented:
@wjwwood All tests passing.

wjwwood commented:
Yup, I was just looking over it again. LGTM

wjwwood commented:
:tada:
